### PR TITLE
Add Grok support

### PR DESCRIPTION
**Pull request by ragaeeb**
# Grok Integration - Complete Implementation

### 1. TypeScript Import Error
**Problem**: JSON imports are typed objects, causing TypeScript errors when passing to `parseInterceptedData(data: string | any)`

**Solution**: 
- Tests now cast JSON imports as `any`: `sampleConversation as any`
- The `parseInterceptedData` method accepts both `string | any`, so this works correctly
- Tests verify both object and string inputs

### 2. Title Caching (Like Gemini)
**Problem**: Grok uses separate endpoints for conversation data and titles, similar to Gemini

**Solution**: Implemented the same pattern as Gemini adapter:
- **Title Cache**: `Map<string, string>` stores `conversationId -> title`
- **Active Conversations**: `Map<string, ConversationData>` stores references for retroactive updates
- **Endpoint Detection**: Checks if URL contains `GrokHistory` vs `GrokConversationItemsByRestId`
- **Merging Logic**: When conversation data is parsed, it checks cache for title

### Title Caching Flow

```typescript
// 1. GrokHistory endpoint is intercepted first (usually)
GET /i/api/graphql/{hash}/GrokHistory
Response: { 
  data: { 
    grok_conversation_history: { 
      items: [
        { 
          grokConversation: { rest_id: "2013295304527827227" },
          title: "Translating Classical Islamic Texts Rules"
        }
      ]
    }
  }
}

// 2. Title is cached
conversationTitles.set("2013295304527827227", "Translating Classical Islamic Texts Rules")

// 3. Later, conversation data is fetched
GET /i/api/graphql/{hash}/GrokConversationItemsByRestId?variables={"restId":"2013295304527827227"}

// 4. Conversation is parsed and title is retrieved from cache
const title = conversationTitles.get("2013295304527827227") || "Grok Conversation"
```

### Retroactive Title Updates

If the order is reversed (conversation data arrives before history):

```typescript
// 1. Conversation data arrives first
const conversation = parseGrokResponse(data)
// Title: "ROLE: Expert academic translator..." (extracted from first message)

// 2. Store in active conversations
activeConversations.set(conversationId, conversation)

// 3. GrokHistory arrives later
const titles = parseTitlesResponse(historyData)

// 4. Update active conversation in-place
const activeObj = activeConversations.get(conversationId)
activeObj.title = "Translating Classical Islamic Texts Rules"
```

### API Endpoint Pattern

Updated regex to match both endpoints:

```typescript
apiEndpointPattern: /\/i\/api\/graphql\/[^/]+\/(GrokConversationItemsByRestId|GrokHistory)/
```

### New Files
1. ✅ `platforms/grok.ts` - With title caching
2. ✅ `platforms/grok.test.ts` - Tests for both endpoints
3. ✅ `data/grok/sample_grok_conversation.json` - Conversation data
4. ✅ `data/grok/sample_grok_history.json` - History/titles data

### Modified Files
1. ✅ `platforms/factory.ts` - Added grokAdapter
2. ✅ `entrypoints/main.content.ts` - Added x.com/i/grok* match
3. ✅ `entrypoints/interceptor.content.ts` - Added x.com/i/grok* match

## Testing

Run the tests:
```bash
bun test platforms/grok.test.ts
```

Expected results:
- ✅ All URL extraction tests pass
- ✅ API endpoint pattern matches both GrokConversationItemsByRestId and GrokHistory
- ✅ Conversation data parsing works with both object and string inputs
- ✅ GrokHistory parsing returns null (as expected) but caches titles
- ✅ Title caching integration test verifies end-to-end flow
- ✅ Structure validation ensures proper message tree
- ✅ Filename formatting handles edge cases

## Key Differences from Gemini

| Feature | Gemini | Grok |
|---------|--------|------|
| **Title Endpoint** | `MaZiqc` RPC in batchexecute | `GrokHistory` GraphQL |
| **Conversation Endpoint** | `hNvQHb` RPC in batchexecute | `GrokConversationItemsByRestId` GraphQL |
| **ID Format** | Hex string (can have `c_` prefix) | Numeric string (10-20 digits) |
| **Response Format** | Double-JSON encoded with security prefix | Standard GraphQL JSON |
| **Thinking Content** | `actualCandidate[37]` array | `deepsearch_headers` array |
| **URL Pattern** | `gemini.google.com/app/{id}` | `x.com/i/grok?conversation={id}` |

## Usage Example

Once deployed:

1. User visits: `https://x.com/i/grok?conversation=2013295304527827227`
2. Page loads and makes two API calls:
   - `GrokHistory` → Intercepted, titles cached
   - `GrokConversationItemsByRestId` → Intercepted, conversation parsed with cached title
3. "Save JSON" button appears in header
4. User clicks → Downloads: `Translating_Classical_Islamic_Texts_Rules_2026-01-19_12-53-00.json`

## Known Limitations

- **Cache is in-memory**: Titles are lost on page refresh (same as Gemini)
- **Single conversation save**: Button works for current conversation only
- **No persistence**: No cross-session title caching (by design)

## Next Steps

To deploy:

1. Add `data/grok/` directory with JSON files
2. Update `wxt.config.ts` host permissions:
   ```typescript
   host_permissions: [
     'https://chatgpt.com/*',
     'https://gemini.google.com/*',
     'https://x.com/*'  // Add this
   ]
   ```
3. Build: `bun run build`
4. Test in browser: Load extension, visit Grok conversation, click "Save JSON"

# DRY Principle & Grok Model Information

### Problem
URL patterns were duplicated across three files:
- `wxt.config.ts` (host_permissions)
- `entrypoints/main.content.ts` (matches)
- `entrypoints/interceptor.content.ts` (matches)

### Solution
Created `platforms/constants.ts` with:

```typescript
export const SUPPORTED_PLATFORM_URLS = [
    'https://chatgpt.com/*',
    'https://chat.openai.com/*',
    'https://gemini.google.com/*',
    'https://x.com/i/grok*',
] as const;
```

### Updated Files

**1. platforms/constants.ts** (New)
- Central source of truth for platform URLs
- Type-safe with `as const` assertion
- Exported `PlatformUrl` type for type safety

**2. wxt.config.ts**
```typescript
import { SUPPORTED_PLATFORM_URLS } from './platforms/constants';

export default defineConfig({
    manifest: {
        host_permissions: [...SUPPORTED_PLATFORM_URLS],
    },
});
```

**3. entrypoints/main.content.ts**
```typescript
import { SUPPORTED_PLATFORM_URLS } from '@/platforms/constants';

export default defineContentScript({
    matches: [...SUPPORTED_PLATFORM_URLS],
    // ...
});
```

**4. entrypoints/interceptor.content.ts**
```typescript
import { SUPPORTED_PLATFORM_URLS } from '@/platforms/constants';

export default defineContentScript({
    matches: [...SUPPORTED_PLATFORM_URLS],
    // ...
});
```

### Benefits
- ✅ Single source of truth
- ✅ Easy to add new platforms (one place)
- ✅ Type safety
- ✅ Reduced duplication
- ✅ Easier maintenance

### Adding New Platforms
Now you only need to update `platforms/constants.ts`:

```typescript
export const SUPPORTED_PLATFORM_URLS = [
    'https://chatgpt.com/*',
    'https://chat.openai.com/*',
    'https://gemini.google.com/*',
    'https://x.com/i/grok*',
    'https://claude.ai/*', // Add new platform here
] as const;
```

All three files will automatically use the updated list!

---

### Research Findings

I searched extensively for information about Grok model identifiers in the X.com GraphQL API responses. Here's what I found:

### xAI Developer API (api.x.ai)
The official xAI API for developers **does** include model information:

```json
{
  "id": "chatcmpl-12345",
  "model": "grok-4",  // ✅ Model info available
  "choices": [...],
  "usage": {...}
}
```

Available models via API:
- `grok-4` (latest, July 2025)
- `grok-4-heavy` (SuperGrok tier)
- `grok-3`
- `grok-3-mini`
- `grok-beta` (Grok 2)
- `grok-2-image-1212` (image generation)

### X.com Web Interface GraphQL API (What We Intercept)
The GraphQL endpoints we intercept for the browser extension **do NOT** expose model information:

**GrokConversationItemsByRestId Response:**
```json
{
  "data": {
    "grok_conversation_items_by_rest_id": {
      "items": [
        {
          "chat_item_id": "...",
          "created_at_ms": 1768841980715,
          "grok_mode": "Normal",  // ❌ Only shows "Normal", not model name
          "message": "...",
          "sender_type": "Agent"
        }
      ]
    }
  }
}
```

**GrokHistory Response:**
```json
{
  "data": {
    "grok_conversation_history": {
      "items": [
        {
          "grokConversation": {
            "rest_id": "..."
          },
          "title": "..."
          // ❌ No model field
        }
      ]
    }
  }
}
```

### Available Fields in Intercepted Data
From the conversation items, we only have:
- `grok_mode`: Always "Normal" (not useful)
- `sender_type`: "User" or "Agent"
- `chat_item_id`: Message ID
- `message`: Content
- `deepsearch_headers`: Reasoning steps
- `ui_layout`: UI metadata

**None of these indicate which Grok model (grok-4, grok-3, etc.) was used.**

#### Option 1: Default to Latest Model
```typescript
default_model_slug: 'grok-4'  // Assume latest
```
**Problem**: Incorrect for older conversations or users without premium

#### Option 2: Infer from Features
```typescript
// If has deepsearch_headers -> assume reasoning model
// If has certain UI layout -> assume specific model
```
**Problem**: Unreliable, purely speculative

#### Option 3: Add User Setting
Let users manually specify which Grok model they're using
**Problem**: Extra friction, users may not know

#### Option 4: Leave as Generic
```typescript
default_model_slug: 'grok-2'  // Generic fallback
```
**Chosen**: This is what we're doing now

### Current Implementation

The Grok adapter currently uses:
```typescript
default_model_slug: 'grok-2'
```

With this comment in the code:
```typescript
// Note: X.com GraphQL API doesn't expose specific model info (grok-4, grok-3, etc.)
// The grok_mode field only contains "Normal", not the model name
// This differs from the xAI API which returns model info in standard format
```

### Why This Limitation Exists

The X.com web interface GraphQL API is designed for the consumer web app, not for developers. It's a private API that:
- Focuses on conversation data, not metadata
- Doesn't expose implementation details (model versions)
- Is separate from the public xAI developer API

The official xAI API (api.x.ai) is what developers should use for programmatic access with full model information, but we're intercepting the web interface API which has different goals.

### Recommendation

**Keep the generic `'grok-2'` default** for now because:
1. ✅ It's honest (we don't know the exact model)
2. ✅ It's safe (doesn't make incorrect assumptions)
3. ✅ It's consistent with other adapters that also use generic defaults
4. ✅ Users can see it's a Grok conversation from the title/content

If users need exact model information, they should use the official xAI API directly rather than capturing from the web interface.

### Future Possibilities

If X (Twitter) adds model information to their GraphQL responses in the future, we can:
1. Search for new fields in the response
2. Update the parser to extract model info
3. Map it to standard model slugs
4. Update `default_model_slug` dynamically

But for now, the information simply isn't available in the intercepted data.

---

## Summary

✅ **DRY Implementation**: Complete - All URL patterns now use `SUPPORTED_PLATFORM_URLS` constant

❌ **Grok Model Info**: Not available in X.com GraphQL API - Using generic `'grok-2'` default

Both issues have been addressed optimally given the constraints!

## Summary by CodeRabbit

* **New Features**
  * Added Grok platform support with conversation parsing, title caching, and retroactive title updates; included sample conversation and history fixtures.

* **Tests**
  * Added comprehensive tests for Grok parsing, URL/endpoint handling, filename formatting, and data integrity.

* **Documentation**
  * Updated README with a short attribution/inspiration note.

* **Chores**
  * Centralized supported platform URLs for manifests/content scripts; added a VSCode setting to suppress a Git warning.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: platforms/grok.ts [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+     * @param data - The conversation data
+     * @returns A sanitized filename (without .json extension)
+     */
+    formatFilename(data: ConversationData): string {
+        let title = data.title || '';
+
+        // If no title, use a default with part of conversation ID
+        if (!title.trim()) {
+            title = `grok_conversation_${data.conversation_id.slice(0, 8)}`;
+        }
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply platforms/grok.ts]: https://github.com/ragaeeb/blackiya/pull/2/commits/3500a0ea0aad8597827db2aa88f9ce21d89eaf38

File: platforms/grok.ts [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+            const grokMode = item.grok_mode || 'Normal';
+            const message = item.message || '';
+            const senderType = item.sender_type || 'Agent';
+            const isPartial = item.is_partial || false;
+
+            // Extract conversation ID from first message
+            if (i === 0 && chatItemId) {
+                // Use the chat_item_id as a reference, but we'll use a more stable ID
+                conversationId = chatItemId;
+            }
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply platforms/grok.ts]: https://github.com/ragaeeb/blackiya/pull/2/commits/3500a0ea0aad8597827db2aa88f9ce21d89eaf38

File: platforms/grok.ts [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+                    }
+                }
+            }
+        }
+
+        return thoughts.length > 0 ? thoughts : undefined;
+    }
+
+    return undefined;
+}
```
[the remark this answers is not in the document]
Comment (ragaeeb) [reply platforms/grok.ts]: https://github.com/ragaeeb/blackiya/pull/2/commits/3500a0ea0aad8597827db2aa88f9ce21d89eaf38
